Error while checking build/O2/fullCI for adf3e56d0d38bafb9b6bef8a1433c905f7b115f8 at 2022-10-31 15:42:
```
## sw/BUILD/o2checkcode-latest/log
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/Common/test/testGPUsortCUDA.cu:22:10: error: 'boost/test/unit_test.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstruction.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstructionCPU.cxx:45:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUTracking/display/GPUDisplay.cxx:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDAGenRTC.cu:16:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:8: error: unknown type name '__host__' [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:27: error: expected ';' after top level declarator [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:33: error: overloaded 'operator+=' must have at least one parameter of class or enumeration type [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:44: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:53: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:85:1: error: unknown type name '__global__' [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:19: error: use of undeclared identifier 'blockIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:32: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:45: error: use of undeclared identifier 'threadIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:78: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/GPU/GPUbenchmark/hip/benchmark.hip.cxx:199:35: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10186-slc8_x86-64/0/Detectors/TOF/calibration/src/TOFChannelCalibrator.cxx:23:10: error: 'omp.h' file not found [clang-diagnostic-error]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10186/adf3e56d0d38bafb9b6bef8a1433c905f7b115f8/build_O2_fullCI/pretty.html).

--- a/Detectors/ITSMFT/ITS/tracking/include/ITStracking/Configuration.h
+++ b/Detectors/ITSMFT/ITS/tracking/include/ITStracking/Configuration.h
@@ -106,6 +106,8 @@ struct VertexingParameters {
   int maxTrackletsPerCluster = 2e3;
   int phiSpan = -1;
   int zSpan = -1;
+
+  int nThreads = 1;
 };
 
 struct VertexerHistogramsConfiguration {

--- a/Detectors/ITSMFT/ITS/tracking/include/ITStracking/TimeFrame.h
+++ b/Detectors/ITSMFT/ITS/tracking/include/ITStracking/TimeFrame.h
@@ -372,14 +372,14 @@ inline gsl::span<int> TimeFrame::getIndexTable(int rofId, int layer)
           static_cast<gsl::span<int>::size_type>(mIndexTableUtils.getNphiBins() * mIndexTableUtils.getNzBins() + 1)};
 }
 
-inline std::vector<Line>& TimeFrame::getLines(int tf)
+inline std::vector<Line>& TimeFrame::getLines(int rof)
 {
-  return mLines[tf];
+  return mLines[rof];
 }
 
-inline std::vector<ClusterLines>& TimeFrame::getTrackletClusters(int tf)
+inline std::vector<ClusterLines>& TimeFrame::getTrackletClusters(int rof)
 {
-  return mTrackletClusters[tf];
+  return mTrackletClusters[rof];
 }
 
 template <typename... T>

--- a/Detectors/ITSMFT/ITS/tracking/include/ITStracking/TimeFrame.h
+++ b/Detectors/ITSMFT/ITS/tracking/include/ITStracking/TimeFrame.h
@@ -157,13 +157,16 @@ class TimeFrame
   int getROfCutAllMult() const { return mCutClusterMult + mCutVertexMult; }
 
   // Vertexer
-  void computeTrackletsScans();
+  void computeTrackletsScans(const int nThreads = 1);
   int& getNTrackletsROf(int tf, int combId);
   std::vector<Line>& getLines(int tf);
   std::vector<ClusterLines>& getTrackletClusters(int tf);
   gsl::span<const Tracklet> getFoundTracklets(int rofId, int combId) const;
+  gsl::span<Tracklet> getFoundTracklets(int rofId, int combId);
   gsl::span<const MCCompLabel> getLabelsFoundTracklets(int rofId, int combId) const;
   gsl::span<int> getNTrackletsCluster(int rofId, int combId);
+  uint32_t getTotalTrackletsTF(const int iLayer) { return mTotalTracklets[iLayer]; }
+
   // \Vertexer
 
   void initialiseRoadLabels();
@@ -244,6 +247,7 @@ class TimeFrame
   std::vector<std::vector<int>> mTrackletsIndexROf;
   std::vector<std::vector<MCCompLabel>> mLinesLabels;
   std::vector<std::vector<MCCompLabel>> mVerticesLabels;
+  std::array<uint32_t, 2> mTotalTracklets = {0, 0};
   // \Vertexer
 };
 
@@ -443,9 +447,9 @@ inline gsl::span<int> TimeFrame::getNTrackletsCluster(int rofId, int combId)
   return {&mNTrackletsPerCluster[combId][startIdx], static_cast<gsl::span<int>::size_type>(mROframesClusters[1][rofId + 1] - startIdx)};
 }
 
-inline int& TimeFrame::getNTrackletsROf(int tf, int combId)
+inline int& TimeFrame::getNTrackletsROf(int rof, int combId)
 {
-  return mNTrackletsPerROf[combId][tf];
+  return mNTrackletsPerROf[combId][rof];
 }
 
 inline bool TimeFrame::isRoadFake(int i) const
@@ -476,6 +480,15 @@ inline std::vector<std::vector<std::vector<int>>>& TimeFrame::getCellsNeighbours
 }
 
 inline std::vector<Road>& TimeFrame::getRoads() { return mRoads; }
+
+inline gsl::span<Tracklet> TimeFrame::getFoundTracklets(int rofId, int combId)
+{
+  if (rofId < 0 || rofId >= mNrof) {
+    return gsl::span<Tracklet>();
+  }
+  auto startIdx{mNTrackletsPerROf[combId][rofId]};
+  return {&mTracklets[combId][startIdx], static_cast<gsl::span<Tracklet>::size_type>(mNTrackletsPerROf[combId][rofId + 1] - startIdx)};
+}
 
 inline gsl::span<const Tracklet> TimeFrame::getFoundTracklets(int rofId, int combId) const
 {

--- a/Detectors/ITSMFT/ITS/tracking/include/ITStracking/TrackingConfigParam.h
+++ b/Detectors/ITSMFT/ITS/tracking/include/ITStracking/TrackingConfigParam.h
@@ -34,6 +34,8 @@ struct VertexerParamConfig : public o2::conf::ConfigurableParamHelper<VertexerPa
   int phiSpan = -1;
   int zSpan = -1;
 
+  int nThreads = 1;
+
   O2ParamDef(VertexerParamConfig, "ITSVertexerParam");
 };
 

--- a/Detectors/ITSMFT/ITS/tracking/include/ITStracking/Vertexer.h
+++ b/Detectors/ITSMFT/ITS/tracking/include/ITStracking/Vertexer.h
@@ -53,12 +53,10 @@ class Vertexer
   Vertexer& operator=(const Vertexer&) = delete;
 
   void adoptTimeFrame(TimeFrame& tf);
-  void setROframe(const uint32_t ROframe) { mROframe = ROframe; }
   void setParameters(const VertexingParameters& verPar);
   void getGlobalConfiguration();
   VertexingParameters getVertParameters() const;
 
-  uint32_t getROFrame() const { return mROframe; }
   std::vector<Vertex> exportVertices();
   VertexerTraits* getTraits() const { return mTraits; };
 
@@ -82,9 +80,10 @@ class Vertexer
   void dumpTraits();
   template <typename... T>
   float evaluateTask(void (Vertexer::*)(T...), const char*, std::function<void(std::string s)> logger, T&&... args);
+  void printEpilog(std::function<void(std::string s)> logger, const float total);
 
  private:
-  std::uint32_t mROframe = 0;
+  std::uint32_t mTimeFrameCounter = 0;
 
   VertexerTraits* mTraits = nullptr; /// Observer pointer, not owned by this class
   TimeFrame* mTimeFrame = nullptr;   /// Observer pointer, not owned by this class

--- a/Detectors/ITSMFT/ITS/tracking/include/ITStracking/VertexerTraits.h
+++ b/Detectors/ITSMFT/ITS/tracking/include/ITStracking/VertexerTraits.h
@@ -81,9 +81,12 @@ class VertexerTraits
   void setIsGPU(const unsigned char isgpu) { mIsGPU = isgpu; };
   unsigned char getIsGPU() const { return mIsGPU; };
   void dumpVertexerTraits();
+  void setNThreads(int n);
+  int getNThreads() const { return mNThreads; }
 
  protected:
   unsigned char mIsGPU;
+  int mNThreads = 1;
 
   VertexingParameters mVrtParams;
   IndexTableUtils mIndexTableUtils;

--- a/Detectors/ITSMFT/ITS/tracking/src/TimeFrame.cxx
+++ b/Detectors/ITSMFT/ITS/tracking/src/TimeFrame.cxx
@@ -24,6 +24,10 @@
 
 #include <iostream>
 
+#ifdef WITH_OPENMP
+#include <omp.h>
+#endif
+
 namespace
 {
 struct ClusterHelper {
@@ -476,10 +480,13 @@ void TimeFrame::printROFoffsets()
   }
 }
 
-void TimeFrame::computeTrackletsScans()
+void TimeFrame::computeTrackletsScans(const int nThreads)
 {
-  std::exclusive_scan(mNTrackletsPerROf[0].begin(), mNTrackletsPerROf[0].end(), mNTrackletsPerROf[0].begin(), 0);
-  std::exclusive_scan(mNTrackletsPerROf[1].begin(), mNTrackletsPerROf[1].end(), mNTrackletsPerROf[1].begin(), 0);
+#pragma omp parallel for num_threads(nThreads > 1 ? 2 : nThreads)
+  for (ushort iLayer = 0; iLayer < 2; ++iLayer) {
+    mTotalTracklets[iLayer] = std::accumulate(mNTrackletsPerROf[iLayer].begin(), mNTrackletsPerROf[iLayer].end(), 0);
+    std::exclusive_scan(mNTrackletsPerROf[iLayer].begin(), mNTrackletsPerROf[iLayer].end(), mNTrackletsPerROf[iLayer].begin(), 0);
+  }
 }
 
 } // namespace its

--- a/Detectors/ITSMFT/ITS/tracking/src/TimeFrame.cxx
+++ b/Detectors/ITSMFT/ITS/tracking/src/TimeFrame.cxx
@@ -482,7 +482,7 @@ void TimeFrame::printROFoffsets()
 
 void TimeFrame::computeTrackletsScans(const int nThreads)
 {
-#pragma omp parallel for num_threads(nThreads > 1 ? 2 : nThreads)
+  // #pragma omp parallel for num_threads(nThreads > 1 ? 2 : nThreads)
   for (ushort iLayer = 0; iLayer < 2; ++iLayer) {
     mTotalTracklets[iLayer] = std::accumulate(mNTrackletsPerROf[iLayer].begin(), mNTrackletsPerROf[iLayer].end(), 0);
     std::exclusive_scan(mNTrackletsPerROf[iLayer].begin(), mNTrackletsPerROf[iLayer].end(), mNTrackletsPerROf[iLayer].begin(), 0);

--- a/Detectors/ITSMFT/ITS/tracking/src/TrackerTraits.cxx
+++ b/Detectors/ITSMFT/ITS/tracking/src/TrackerTraits.cxx
@@ -998,6 +998,7 @@ void TrackerTraits::setNThreads(int n)
   mNThreads = 1;
 #endif
 }
+
 int TrackerTraits::getTFNumberOfClusters() const
 {
   return mTimeFrame->getNumberOfClusters();

--- a/Detectors/ITSMFT/ITS/tracking/src/Vertexer.cxx
+++ b/Detectors/ITSMFT/ITS/tracking/src/Vertexer.cxx
@@ -41,10 +41,10 @@ float Vertexer::clustersToVertices(std::function<void(std::string s)> logger)
   float total{0.f};
   TrackingParameters trkPars;
   total += evaluateTask(&Vertexer::initialiseVertexer, "Vertexer initialisation", logger, trkPars);
-  total += evaluateTask(&Vertexer::findTracklets, "Tracklet finding", logger);
-  total += evaluateTask(&Vertexer::validateTracklets, "Adjacent tracklets validation", logger);
-  total += evaluateTask(&Vertexer::findVertices, "Vertex finding", logger);
-
+  total += evaluateTask(&Vertexer::findTracklets, "Vertexer tracklet finding", logger);
+  total += evaluateTask(&Vertexer::validateTracklets, "Vertexer adjacent tracklets validation", logger);
+  total += evaluateTask(&Vertexer::findVertices, "Vertexer vertex finding", logger);
+  printEpilog(logger, total);
   return total;
 }
 
@@ -67,6 +67,7 @@ void Vertexer::getGlobalConfiguration()
   verPar.clusterContributorsCut = vc.clusterContributorsCut;
   verPar.maxTrackletsPerCluster = vc.maxTrackletsPerCluster;
   verPar.phiSpan = vc.phiSpan;
+  verPar.nThreads = vc.nThreads;
 
   mTraits->updateVertexingParameters(verPar);
 }
@@ -76,5 +77,11 @@ void Vertexer::adoptTimeFrame(TimeFrame& tf)
   mTimeFrame = &tf;
   mTraits->adoptTimeFrame(&tf);
 }
+
+void Vertexer::printEpilog(std::function<void(std::string s)> logger, const float total)
+{
+  logger(fmt::format(" - Timeframe {} vertexing completed in: {}ms, using {} thread(s).", mTimeFrameCounter++, total, mTraits->getNThreads()));
+}
+
 } // namespace its
 } // namespace o2

--- a/Detectors/ITSMFT/ITS/tracking/src/VertexerTraits.cxx
+++ b/Detectors/ITSMFT/ITS/tracking/src/VertexerTraits.cxx
@@ -161,62 +161,68 @@ void VertexerTraits::updateVertexingParameters(const VertexingParameters& vrtPar
 
 void VertexerTraits::computeTracklets()
 {
-#pragma omp parallel for num_threads(mNThreads)
-  for (int rofId = 0; rofId < mTimeFrame->getNrof(); ++rofId) {
-    trackleterKernelHost<TrackletMode::Layer0Layer1, true>(
-      mTimeFrame->getClustersOnLayer(rofId, 0),
-      mTimeFrame->getClustersOnLayer(rofId, 1),
-      mTimeFrame->getIndexTable(rofId, 0).data(),
-      mVrtParams.phiCut,
-      mTimeFrame->getTracklets()[0],
-      mTimeFrame->getNTrackletsCluster(rofId, 0),
-      mIndexTableUtils,
-      rofId,
-      0,
-      mVrtParams.maxTrackletsPerCluster);
-    trackleterKernelHost<TrackletMode::Layer1Layer2, true>(
-      mTimeFrame->getClustersOnLayer(rofId, 2),
-      mTimeFrame->getClustersOnLayer(rofId, 1),
-      mTimeFrame->getIndexTable(rofId, 2).data(),
-      mVrtParams.phiCut,
-      mTimeFrame->getTracklets()[1],
-      mTimeFrame->getNTrackletsCluster(rofId, 1),
-      mIndexTableUtils,
-      rofId,
-      0,
-      mVrtParams.maxTrackletsPerCluster);
-    mTimeFrame->getNTrackletsROf(rofId, 0) = std::accumulate(mTimeFrame->getNTrackletsCluster(rofId, 0).begin(), mTimeFrame->getNTrackletsCluster(rofId, 0).end(), 0);
-    mTimeFrame->getNTrackletsROf(rofId, 1) = std::accumulate(mTimeFrame->getNTrackletsCluster(rofId, 1).begin(), mTimeFrame->getNTrackletsCluster(rofId, 1).end(), 0);
-  }
 
-  mTimeFrame->computeTrackletsScans(mNThreads);
-  mTimeFrame->getTracklets()[0].resize(mTimeFrame->getTotalTrackletsTF(0));
-  mTimeFrame->getTracklets()[1].resize(mTimeFrame->getTotalTrackletsTF(1));
+#pragma omp parallel num_threads(mNThreads)
+  {
+#pragma omp for schedule(dynamic)
+    for (int rofId = 0; rofId < mTimeFrame->getNrof(); ++rofId) {
+      trackleterKernelHost<TrackletMode::Layer0Layer1, true>(
+        mTimeFrame->getClustersOnLayer(rofId, 0),
+        mTimeFrame->getClustersOnLayer(rofId, 1),
+        mTimeFrame->getIndexTable(rofId, 0).data(),
+        mVrtParams.phiCut,
+        mTimeFrame->getTracklets()[0],
+        mTimeFrame->getNTrackletsCluster(rofId, 0),
+        mIndexTableUtils,
+        rofId,
+        0,
+        mVrtParams.maxTrackletsPerCluster);
+      trackleterKernelHost<TrackletMode::Layer1Layer2, true>(
+        mTimeFrame->getClustersOnLayer(rofId, 2),
+        mTimeFrame->getClustersOnLayer(rofId, 1),
+        mTimeFrame->getIndexTable(rofId, 2).data(),
+        mVrtParams.phiCut,
+        mTimeFrame->getTracklets()[1],
+        mTimeFrame->getNTrackletsCluster(rofId, 1),
+        mIndexTableUtils,
+        rofId,
+        0,
+        mVrtParams.maxTrackletsPerCluster);
+      mTimeFrame->getNTrackletsROf(rofId, 0) = std::accumulate(mTimeFrame->getNTrackletsCluster(rofId, 0).begin(), mTimeFrame->getNTrackletsCluster(rofId, 0).end(), 0);
+      mTimeFrame->getNTrackletsROf(rofId, 1) = std::accumulate(mTimeFrame->getNTrackletsCluster(rofId, 1).begin(), mTimeFrame->getNTrackletsCluster(rofId, 1).end(), 0);
+    }
+#pragma omp single
+    mTimeFrame->computeTrackletsScans(mNThreads);
+#pragma omp single
+    mTimeFrame->getTracklets()[0].resize(mTimeFrame->getTotalTrackletsTF(0));
+#pragma omp single
+    mTimeFrame->getTracklets()[1].resize(mTimeFrame->getTotalTrackletsTF(1));
 
-#pragma omp parallel for num_threads(mNThreads)
-  for (int rofId = 0; rofId < mTimeFrame->getNrof(); ++rofId) {
-    trackleterKernelHost<TrackletMode::Layer0Layer1, false>(
-      mTimeFrame->getClustersOnLayer(rofId, 0),
-      mTimeFrame->getClustersOnLayer(rofId, 1),
-      mTimeFrame->getIndexTable(rofId, 0).data(),
-      mVrtParams.phiCut,
-      mTimeFrame->getTracklets()[0],
-      mTimeFrame->getNTrackletsCluster(rofId, 0),
-      mIndexTableUtils,
-      rofId,
-      mTimeFrame->getNTrackletsROf(rofId, 0),
-      mVrtParams.maxTrackletsPerCluster);
-    trackleterKernelHost<TrackletMode::Layer1Layer2, false>(
-      mTimeFrame->getClustersOnLayer(rofId, 2),
-      mTimeFrame->getClustersOnLayer(rofId, 1),
-      mTimeFrame->getIndexTable(rofId, 2).data(),
-      mVrtParams.phiCut,
-      mTimeFrame->getTracklets()[1],
-      mTimeFrame->getNTrackletsCluster(rofId, 1),
-      mIndexTableUtils,
-      rofId,
-      mTimeFrame->getNTrackletsROf(rofId, 1),
-      mVrtParams.maxTrackletsPerCluster);
+#pragma omp for schedule(dynamic)
+    for (int rofId = 0; rofId < mTimeFrame->getNrof(); ++rofId) {
+      trackleterKernelHost<TrackletMode::Layer0Layer1, false>(
+        mTimeFrame->getClustersOnLayer(rofId, 0),
+        mTimeFrame->getClustersOnLayer(rofId, 1),
+        mTimeFrame->getIndexTable(rofId, 0).data(),
+        mVrtParams.phiCut,
+        mTimeFrame->getTracklets()[0],
+        mTimeFrame->getNTrackletsCluster(rofId, 0),
+        mIndexTableUtils,
+        rofId,
+        mTimeFrame->getNTrackletsROf(rofId, 0),
+        mVrtParams.maxTrackletsPerCluster);
+      trackleterKernelHost<TrackletMode::Layer1Layer2, false>(
+        mTimeFrame->getClustersOnLayer(rofId, 2),
+        mTimeFrame->getClustersOnLayer(rofId, 1),
+        mTimeFrame->getIndexTable(rofId, 2).data(),
+        mVrtParams.phiCut,
+        mTimeFrame->getTracklets()[1],
+        mTimeFrame->getNTrackletsCluster(rofId, 1),
+        mIndexTableUtils,
+        rofId,
+        mTimeFrame->getNTrackletsROf(rofId, 1),
+        mVrtParams.maxTrackletsPerCluster);
+    }
   }
 
   /// Create tracklets labels for L0-L1, information is as flat as in tracklets vector (no rofId)
@@ -276,11 +282,13 @@ void VertexerTraits::computeTracklets()
   out01.close();
   out12.close();
 #endif
-}
+} // namespace its
 
 void VertexerTraits::computeTrackletMatching()
 {
-  for (int rofId{0}; rofId < mTimeFrame->getNrof(); ++rofId) {
+#pragma omp parallel for num_threads(mNThreads) schedule(dynamic)
+  for (int rofId = 0; rofId < mTimeFrame->getNrof(); ++rofId) {
+    mTimeFrame->getLines(rofId).reserve(mTimeFrame->getNTrackletsCluster(rofId, 0).size());
     trackletSelectionKernelHost(
       mTimeFrame->getClustersOnLayer(rofId, 0),
       mTimeFrame->getClustersOnLayer(rofId, 1),

--- a/Detectors/ITSMFT/ITS/tracking/src/VertexerTraits.cxx
+++ b/Detectors/ITSMFT/ITS/tracking/src/VertexerTraits.cxx
@@ -24,6 +24,10 @@
 #include <ostream>
 #endif
 
+#ifdef WITH_OPENMP
+#include <omp.h>
+#endif
+
 namespace o2
 {
 namespace its
@@ -31,22 +35,24 @@ namespace its
 using boost::histogram::indexed;
 using constants::math::TwoPi;
 
-template <TrackletMode Mode>
-void trackleterKernelSerial(
+template <TrackletMode Mode, bool DryRun>
+void trackleterKernelHost(
   const gsl::span<const Cluster>& clustersNextLayer,    // 0 2
   const gsl::span<const Cluster>& clustersCurrentLayer, // 1 1
   int* indexTableNext,
   const float phiCut,
-  std::vector<Tracklet>& Tracklets,
+  std::vector<Tracklet>& tracklets,
   gsl::span<int> foundTracklets,
   const IndexTableUtils& utils,
   const int rof,
+  const int rofFoundTrackletsOffset,
   const int maxTrackletsPerCluster = static_cast<int>(2e3))
 {
   const int PhiBins{utils.getNphiBins()};
   const int ZBins{utils.getNzBins()};
   // loop on layer1 clusters
-  for (unsigned int iCurrentLayerClusterIndex{0}; iCurrentLayerClusterIndex < clustersCurrentLayer.size(); ++iCurrentLayerClusterIndex) {
+  int cumulativeStoredTracklets{0};
+  for (int iCurrentLayerClusterIndex = 0; iCurrentLayerClusterIndex < clustersCurrentLayer.size(); ++iCurrentLayerClusterIndex) {
     int storedTracklets{0};
     const Cluster& currentCluster{clustersCurrentLayer[iCurrentLayerClusterIndex]};
     const int4 selectedBinsRect{VertexerTraits::getBinsRect(currentCluster, (int)Mode, 0.f, 50.f, phiCut / 2, utils)};
@@ -65,10 +71,12 @@ void trackleterKernelSerial(
           const Cluster& nextCluster{clustersNextLayer[iNextLayerClusterIndex]};
           if (o2::gpu::GPUCommonMath::Abs(currentCluster.phi - nextCluster.phi) < phiCut) {
             if (storedTracklets < maxTrackletsPerCluster) {
-              if constexpr (Mode == TrackletMode::Layer0Layer1) {
-                Tracklets.emplace_back(iNextLayerClusterIndex, iCurrentLayerClusterIndex, nextCluster, currentCluster, rof, rof);
-              } else {
-                Tracklets.emplace_back(iCurrentLayerClusterIndex, iNextLayerClusterIndex, currentCluster, nextCluster, rof, rof);
+              if constexpr (!DryRun) {
+                if constexpr (Mode == TrackletMode::Layer0Layer1) {
+                  tracklets[rofFoundTrackletsOffset + cumulativeStoredTracklets + storedTracklets] = Tracklet{iNextLayerClusterIndex, iCurrentLayerClusterIndex, nextCluster, currentCluster, rof, rof};
+                } else {
+                  tracklets[rofFoundTrackletsOffset + cumulativeStoredTracklets + storedTracklets] = Tracklet{iCurrentLayerClusterIndex, iNextLayerClusterIndex, currentCluster, nextCluster, rof, rof};
+                }
               }
               ++storedTracklets;
             }
@@ -76,11 +84,15 @@ void trackleterKernelSerial(
         }
       }
     }
-    foundTracklets[iCurrentLayerClusterIndex] = storedTracklets;
+    if constexpr (DryRun) {
+      foundTracklets[iCurrentLayerClusterIndex] = storedTracklets;
+    } else {
+      cumulativeStoredTracklets += storedTracklets;
+    }
   }
 }
 
-void trackletSelectionKernelSerial(
+void trackletSelectionKernelHost(
   const gsl::span<const Cluster> clusters0, // 0
   const gsl::span<const Cluster> clusters1, // 1
   const gsl::span<const Tracklet>& tracklets01,
@@ -144,12 +156,14 @@ void VertexerTraits::updateVertexingParameters(const VertexingParameters& vrtPar
   mVrtParams.phiSpan = static_cast<int>(std::ceil(mIndexTableUtils.getNphiBins() * mVrtParams.phiCut /
                                                   constants::math::TwoPi));
   mVrtParams.zSpan = static_cast<int>(std::ceil(mVrtParams.zCut * mIndexTableUtils.getInverseZCoordinate(0)));
+  setNThreads(mVrtParams.nThreads);
 }
 
 void VertexerTraits::computeTracklets()
 {
-  for (int rofId{0}; rofId < mTimeFrame->getNrof(); ++rofId) {
-    trackleterKernelSerial<TrackletMode::Layer0Layer1>(
+#pragma omp parallel for num_threads(mNThreads)
+  for (int rofId = 0; rofId < mTimeFrame->getNrof(); ++rofId) {
+    trackleterKernelHost<TrackletMode::Layer0Layer1, true>(
       mTimeFrame->getClustersOnLayer(rofId, 0),
       mTimeFrame->getClustersOnLayer(rofId, 1),
       mTimeFrame->getIndexTable(rofId, 0).data(),
@@ -158,8 +172,9 @@ void VertexerTraits::computeTracklets()
       mTimeFrame->getNTrackletsCluster(rofId, 0),
       mIndexTableUtils,
       rofId,
+      0,
       mVrtParams.maxTrackletsPerCluster);
-    trackleterKernelSerial<TrackletMode::Layer1Layer2>(
+    trackleterKernelHost<TrackletMode::Layer1Layer2, true>(
       mTimeFrame->getClustersOnLayer(rofId, 2),
       mTimeFrame->getClustersOnLayer(rofId, 1),
       mTimeFrame->getIndexTable(rofId, 2).data(),
@@ -168,11 +183,41 @@ void VertexerTraits::computeTracklets()
       mTimeFrame->getNTrackletsCluster(rofId, 1),
       mIndexTableUtils,
       rofId,
+      0,
       mVrtParams.maxTrackletsPerCluster);
     mTimeFrame->getNTrackletsROf(rofId, 0) = std::accumulate(mTimeFrame->getNTrackletsCluster(rofId, 0).begin(), mTimeFrame->getNTrackletsCluster(rofId, 0).end(), 0);
     mTimeFrame->getNTrackletsROf(rofId, 1) = std::accumulate(mTimeFrame->getNTrackletsCluster(rofId, 1).begin(), mTimeFrame->getNTrackletsCluster(rofId, 1).end(), 0);
   }
-  mTimeFrame->computeTrackletsScans();
+
+  mTimeFrame->computeTrackletsScans(mNThreads);
+  mTimeFrame->getTracklets()[0].reserve(mTimeFrame->getTotalTrackletsTF(0));
+  mTimeFrame->getTracklets()[1].reserve(mTimeFrame->getTotalTrackletsTF(1));
+
+#pragma omp parallel for num_threads(mNThreads)
+  for (int rofId = 0; rofId < mTimeFrame->getNrof(); ++rofId) {
+    trackleterKernelHost<TrackletMode::Layer0Layer1, false>(
+      mTimeFrame->getClustersOnLayer(rofId, 0),
+      mTimeFrame->getClustersOnLayer(rofId, 1),
+      mTimeFrame->getIndexTable(rofId, 0).data(),
+      mVrtParams.phiCut,
+      mTimeFrame->getTracklets()[0],
+      mTimeFrame->getNTrackletsCluster(rofId, 0),
+      mIndexTableUtils,
+      rofId,
+      mTimeFrame->getNTrackletsROf(rofId, 0),
+      mVrtParams.maxTrackletsPerCluster);
+    trackleterKernelHost<TrackletMode::Layer1Layer2, false>(
+      mTimeFrame->getClustersOnLayer(rofId, 2),
+      mTimeFrame->getClustersOnLayer(rofId, 1),
+      mTimeFrame->getIndexTable(rofId, 2).data(),
+      mVrtParams.phiCut,
+      mTimeFrame->getTracklets()[1],
+      mTimeFrame->getNTrackletsCluster(rofId, 1),
+      mIndexTableUtils,
+      rofId,
+      mTimeFrame->getNTrackletsROf(rofId, 1),
+      mVrtParams.maxTrackletsPerCluster);
+  }
 
   /// Create tracklets labels for L0-L1, information is as flat as in tracklets vector (no rofId)
   if (mTimeFrame->hasMCinformation()) {
@@ -236,7 +281,7 @@ void VertexerTraits::computeTracklets()
 void VertexerTraits::computeTrackletMatching()
 {
   for (int rofId{0}; rofId < mTimeFrame->getNrof(); ++rofId) {
-    trackletSelectionKernelSerial(
+    trackletSelectionKernelHost(
       mTimeFrame->getClustersOnLayer(rofId, 0),
       mTimeFrame->getClustersOnLayer(rofId, 1),
       mTimeFrame->getFoundTracklets(rofId, 0),
@@ -408,6 +453,16 @@ void VertexerTraits::computeVertices()
   ln_clus_lines_tree->Write();
   dbg_file->Close();
 #endif
+}
+
+void VertexerTraits::setNThreads(int n)
+{
+#ifdef WITH_OPENMP
+  mNThreads = n > 0 ? n : 1;
+#else
+  mNThreads = 1;
+#endif
+  LOGP(info, "Setting seeding vertexer with {} threads.", mNThreads);
 }
 
 // void VertexerTraits::computeHistVertices()

--- a/Detectors/ITSMFT/ITS/tracking/src/VertexerTraits.cxx
+++ b/Detectors/ITSMFT/ITS/tracking/src/VertexerTraits.cxx
@@ -190,8 +190,8 @@ void VertexerTraits::computeTracklets()
   }
 
   mTimeFrame->computeTrackletsScans(mNThreads);
-  mTimeFrame->getTracklets()[0].reserve(mTimeFrame->getTotalTrackletsTF(0));
-  mTimeFrame->getTracklets()[1].reserve(mTimeFrame->getTotalTrackletsTF(1));
+  mTimeFrame->getTracklets()[0].resize(mTimeFrame->getTotalTrackletsTF(0));
+  mTimeFrame->getTracklets()[1].resize(mTimeFrame->getTotalTrackletsTF(1));
 
 #pragma omp parallel for num_threads(mNThreads)
   for (int rofId = 0; rofId < mTimeFrame->getNrof(); ++rofId) {

